{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c5d91de3",
   "metadata": {},
   "source": [
    "## Example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7617c17",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "sos run ~/codes/xqtl-pipeline/code/intact_pipeline/Padding.ipynb padding \\\n",
    "    --qtl_file `ls /mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/output/susie_per_gene_tad/cache/*rds`  \\\n",
    "    --gwas_path /mnt/mfs/hgrcgrid/homes/hs3393/ADHD/ADHD_finemap/ -J 50 -c ~/test/csg.yml -q csg --job_size 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8947711e",
   "metadata": {},
   "outputs": [],
   "source": [
    "[global]\n",
    "# Workdir\n",
    "parameter: cwd = path(\"output\")\n",
    "# dataset \n",
    "parameter: tissue = ''\n",
    "# QTL data type\n",
    "parameter: QTL = 'eQTL'\n",
    "parameter: GWAS = 'GWAS'\n",
    "parameter: container = ''\n",
    "parameter: entrypoint={('micromamba run -n' + ' ' + container.split('/')[-1][:-4]) if container.endswith('.sif') else f''}\n",
    "parameter: job_size = 1\n",
    "parameter: walltime = \"5h\"\n",
    "parameter: mem = \"8G\"\n",
    "parameter: numThreads = 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bea7c49c-8658-4826-b70e-fa93dd5fbe7f",
   "metadata": {
    "kernel": "SoS",
    "tags": [],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "[padding]\n",
    "parameter: qtl_file = paths\n",
    "parameter: gwas_path = path\n",
    "input: qtl_file, group_by = 1\n",
    "output:  f'{cwd:a}/{QTL}/{QTL}.{_input:bn}.newpadding.rds',f'{cwd:a}/{GWAS}/{GWAS}.{_input:bn}.newpadding.rds'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:nn}.stdout\", stderr = f\"{_output[0]:nn}.stderr\", container = container, entrypoint = entrypoint\n",
    "    suppressPackageStartupMessages(library(tidyverse))\n",
    "    # read all GWAS finemapping result file name to extract the chr, start, end and corresponding file paths  \n",
    "    gwas_finemap_result = list.files(\"${gwas_path}\", full.names=T, pattern = \"\\\\.rds$\")\n",
    "    gwas_file_tb = tibble(file_path = gwas_finemap_result)\n",
    "\n",
    "    # use regex to extract the start, end and chromosome of LD blocks from the file name\n",
    "    match_pattern = function(filename){\n",
    "        pattern <- \"chr[0-9XY]+_(\\\\d+_\\\\d+)\"\n",
    "        result = regmatches(filename, regexpr(pattern, filename))\n",
    "        return(result)\n",
    "    }\n",
    "      \n",
    "    LD_block_position = gwas_file_tb %>% mutate(chr_pos = map(file_path, match_pattern)) %>% \n",
    "      separate(chr_pos, into = c(\"chr\", \"start\", \"end\")) %>% relocate(file_path, .after = end) %>%\n",
    "        mutate(start = as.numeric(start), end = as.numeric(end)) %>% arrange(chr, start) \n",
    "      \n",
    "    # give the absolute path to read RDS of qtl finemapping result rds file  \n",
    "    qtl_file = readRDS(\"${_input}\")$dlpfc_eqtl\n",
    "    # after reading the variants in the rds file, we can use the first variant and last one to extract the chr, start and end of eqtl variants\n",
    "    qtl_chr = regmatches(qtl_file$variable_name[1], regexpr(\"(chr[0-9]+)\", qtl_file$variable_name[1]))\n",
    "    qtl_start  <- as.numeric(sub(\".*:(\\\\d+)_.*\", \"\\\\1\", qtl_file$variable_name[1]))\n",
    "    qtl_end  <- as.numeric(sub(\".*:(\\\\d+)_.*\", \"\\\\1\", qtl_file$variable_name[length(qtl_file$variable_name)]))\n",
    "\n",
    "    # use the regions in qtl to get those LD blocks that overlap with qtl regions\n",
    "    related_LD = LD_block_position %>% filter(chr == qtl_chr) %>% filter((start <= qtl_start & end >= qtl_start) |\n",
    "                                            (start >= qtl_start & end <= qtl_end) | \n",
    "                                            (start <= qtl_end & end >= qtl_end))\n",
    "\n",
    "    # extract those related GWAS finemapping result lbf matrix to form the larger one\n",
    "    cnt = 1\n",
    "    variants = c()\n",
    "    lbf_mtx = list()\n",
    "    for (file in related_LD$file_path){\n",
    "        rds = readRDS(file)\n",
    "        variants = c(variants, rds$variants)\n",
    "        lbf_mtx[[cnt]] = as.data.frame(rds$lbf_variable)\n",
    "        colnames(lbf_mtx[[cnt]]) = rds$variants\n",
    "        cnt = cnt + 1\n",
    "    }\n",
    "\n",
    "    # after combining the matrices, fill those NA with 0 to form the whole matrix\n",
    "    lbf_whole_mtx = bind_rows(lbf_mtx) %>% replace(is.na(.), 0)\n",
    "\n",
    "    # get the shared variants between gwas and qtl\n",
    "\n",
    "    # here is one problem: maybe now the output of finemapping does not need $dlpfc... and the variant name have different format\n",
    "    # chr:9999 or chr_999, so here we change : to _; this brings a lot of trouble\n",
    "    # if now variable name are uniformed, then things are good, remove the str_replcace line\n",
    "      \n",
    "    shared_variant = unlist(intersect(qtl_file$variable_name %>% \n",
    "      map(~ str_replace_all(.x, \":\", \"_\")), colnames(lbf_whole_mtx)))\n",
    "\n",
    "      \n",
    "    # remove those columns that does not share SNP, only keep those shared by two datasetl\n",
    "    GWAS_lbf_matrix = lbf_whole_mtx[, shared_variant]\n",
    "      \n",
    "\n",
    "    # again, change the variant names\n",
    "    colnames(qtl_file$lbf_variable) = unlist(qtl_file$variable_name %>% \n",
    "      map(~ str_replace_all(.x, \":\", \"_\")))\n",
    "\n",
    "    # because in the susie output the cs are recorded by index, so we get the index that are removed\n",
    "    rm_index = which(!(colnames(qtl_file$lbf_variable) %in% shared_variant))\n",
    "      \n",
    "    # also, for qtl data, only keep variants that are shared\n",
    "    qtl_lbf_mtx = qtl_file$lbf_variable[,shared_variant]\n",
    "      \n",
    "\n",
    "    # convert lbf to alpha\n",
    "    lbf_to_alpha_vector = function(lbf, prior_weights = NULL) {\n",
    "      if (is.null(prior_weights)) prior_weights = 1/length(lbf)\n",
    "      maxlbf = max(lbf)\n",
    "      # w is proportional to BF, but subtract max for numerical stability.\n",
    "      w = exp(lbf - maxlbf)\n",
    "      # Posterior prob for each SNP.\n",
    "      w_weighted = w * prior_weights\n",
    "      weighted_sum_w = sum(w_weighted)\n",
    "      alpha = w_weighted / weighted_sum_w\n",
    "      return(alpha)\n",
    "    }\n",
    "\n",
    "    lbf_to_alpha = function(lbf) t(apply(lbf, 1, lbf_to_alpha_vector))\n",
    "\n",
    "    # convert lbf to pip\n",
    "    lbf_to_pip = function(lbf) {\n",
    "        alpha = lbf_to_alpha(lbf)\n",
    "        return(as.vector(1 - apply(1 - alpha,2,prod)))\n",
    "    }\n",
    "\n",
    "\n",
    "    # use new lbf matrix to compute alpha and pip\n",
    "    shared_variable_name<-str_replace(shared_variant, \"_\", \":\")\n",
    "\n",
    "    GWAS_alpha = lbf_to_alpha(GWAS_lbf_matrix)\n",
    "    GWAS_pip = lbf_to_pip(GWAS_lbf_matrix)\n",
    "\n",
    "    qtl_alpha = lbf_to_alpha(qtl_lbf_mtx)\n",
    "    qtl_pip = lbf_to_pip(qtl_lbf_mtx)\n",
    "    colnames(qtl_alpha) =colnames(GWAS_alpha) = names(qtl_pip)  = names(GWAS_pip) = shared_variable_name\n",
    "    # for now, we remove the variants in cs, if that variant is not shared by two traits\n",
    "    ## note: this part may be changed!! now is just a rough strategy\n",
    "    cs_number = length(qtl_file$sets$cs)\n",
    "    new_cs = list()\n",
    "    if(cs_number == 0){\n",
    "        new_cs = NA\n",
    "    }else{\n",
    "        for (i in (1:cs_number)){\n",
    "        new_cs[[i]] = setdiff(qtl_file$sets$cs[[i]], rm_index)\n",
    "        }\n",
    "    }\n",
    "\n",
    "    # the output can be: GWAS+ qtl lbf matrix; alpha; pip\n",
    "    qtl_mu <- qtl_file$mu[,shared_variable_name]\n",
    "    names(qtl_file$X_column_scale_factors) = qtl_file$variable_name\n",
    "    qtl_coef<-colSums(qtl_alpha*qtl_mu)/(qtl_file$X_column_scale_factors[shared_variable_name])\n",
    "\n",
    "\n",
    "    new_qtl_obj <- list( pip =qtl_pip,\n",
    "                        coef = qtl_coef,\n",
    "                        mu = qtl_mu,\n",
    "                        alpha = qtl_alpha,\n",
    "                        sets=list(cs = new_cs),\n",
    "                        variable_name = shared_variable_name\n",
    "    )\n",
    "    names(new_qtl_obj$sets$cs) = names(qtl_file$sets$cs)\n",
    "\n",
    "    new_gwas_obj <- list( pip =GWAS_pip,\n",
    "                        alpha = GWAS_alpha,\n",
    "                        variable_name = shared_variable_name\n",
    "    )\n",
    "    saveRDS(new_qtl_obj,\"${_output[0]}\")\n",
    "    saveRDS(new_gwas_obj,\"${_output[1]}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "version": "0.24.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
