{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "nasty-doubt",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# Feature score from contrast results\n",
    "     From Daniel 5/2023: If the goal is to identify features where we have the strongest significant differences between pairs of cell types, I wonder if we could do some sort of “nested” meta analysis like this:\n",
    "- Do a meta-analysis with the absolute value of  effect sizes (sign doesn’t matter) and standard errors of the pairwise cell type differences within each SNP for a feature - this would give you the meta analysis estimate of the average cell type difference in effect sizes between any pair of cell types\n",
    "- Do a second round of meta-analysis on the SNP-level meta-analyzed effect size and standard error across all the SNPs in the gene - this will give you a meta-analysis estimate of the average cell type difference in the average SNP between any pair of cell types in any SNP\n",
    "\n",
    "That would ultimately yield a single, unsigned Z-score for each feature that could be used to rank the features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "liked-committee",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "import os\n",
    "# Work directory & output directory\n",
    "parameter: cwd = path('./')\n",
    "# The filename prefix for output data\n",
    "parameter: job_size = 1\n",
    "parameter: mem = '60G'\n",
    "parameter: meta_method = \"REML\" #Restricted Maximum Likelihood\n",
    "parameter: container = ''\n",
    "import re\n",
    "parameter: entrypoint= ('micromamba run -a \"\" -n' + ' ' + re.sub(r'(_apptainer:latest|_docker:latest|\\.sif)$', '', container.split('/')[-1])) if container else \"\"\n",
    "#parameter: orig_file = path\n",
    "parameter: analysis_unit = path\n",
    "#parameter:contrast_dir = \"/home/rf2872/Work/MASH_test_csg/MASH_6_celltypes_Dan/contrast/\"\n",
    "##  conditions can be excluded if needs arise. If nothing to exclude keep the default 0\n",
    "import pandas as pd\n",
    "#parameter: analysis_units = path\n",
    "# handle N = per_chunk data-set in one job\n",
    "#parameter: per_chunk = 1\n",
    "regions = [x.replace(\"\\\"\",\"\").strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "\n",
    "analysis_unit = file_target(f\"{cwd:a}/feature_score/posterior_contrast_files\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "portable-acquisition",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# ls the contrast results\n",
    "#[contrast_results]\n",
    "#input: group_by = \"all\"\n",
    "output: f\"{cwd}/feature_score/posterior_contrast_files\"\n",
    "task: trunk_workers = 1, walltime = '2h', trunk_size = 1, mem = '16G', cores = 1, tags = f'{_output:bn}'\n",
    "bash: expand ='${ }', stderr = f\"{_output:n}.stderr\", stdout = f\"{_output:n}.stdout\"\n",
    "     ls ${cwd}/*posterior_contrast > ${_output:r}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dental-float",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# compute feature score from contrast results\n",
    "#[feature_score_meta_1]\n",
    "#regions = [x.replace(\"\\\"\",\"\").strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "parameter: per_chunk = '100'\n",
    "parameter: downsample_ratio = '1'\n",
    "parameter: meta_method = 'REML'\n",
    "parameter: contrast_input = [path(x[0]) for x in regions]\n",
    "input: contrast_input, group_by = per_chunk\n",
    "output: f\"{cwd}/feature_score_meta/cache/mash_posterior_contrast_featurescore{_index+1}.rds\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '20G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    suppressMessages(library(data.table))\n",
    "    suppressMessages(library(tidyverse))\n",
    "    suppressMessages(library(metafor))\n",
    "    out<-NULL\n",
    "    set.seed(999)\n",
    "        for (res in c(${_input:r,})) {\n",
    "            gene <- basename(res) %>%\n",
    "                stringr::str_split(., \"norminal.cis_long_table.\", simplify = TRUE) %>%\n",
    "                .[, 2] %>%\n",
    "                gsub(\"_posterior_contrast.rds\", \"\", .)\n",
    "\n",
    "            tmp_contrast_results <- readRDS(res)\n",
    "            if (${downsample_ratio} < 1) {\n",
    "                # downsample\n",
    "                sample_size <- round(nrow(tmp_contrast_results) * ${downsample_ratio})\n",
    "                sample_rows <- sample(nrow(tmp_contrast_results), size = sample_size)\n",
    "                tmp_contrast_results <- tmp_contrast_results[sample_rows, ]\n",
    "            }\n",
    "\n",
    "            if (tmp_contrast_results %>% select(matches(\"mean_contrast.*deviation\")) %>% ncol() > 0) {\n",
    "                effect_sizes <- tmp_contrast_results %>%\n",
    "                    select(matches(\"mean_contrast.*deviation\")) %>%\n",
    "                    as.matrix()\n",
    "                se_values <- tmp_contrast_results %>%\n",
    "                    select(matches(\"se_contrast.*deviation\")) %>%\n",
    "                    as.matrix()\n",
    "\n",
    "                # Initialize an empty data table to store the feature scores for each condition\n",
    "                feature_scores <- data.table()\n",
    "\n",
    "                # Iterate over each condition\n",
    "                for (i in 1:ncol(effect_sizes)) {\n",
    "                    # Subset the effect sizes and standard errors for the current condition\n",
    "                    effect_sizes_condition <- effect_sizes[, i]\n",
    "                    se_values_condition <- se_values[, i]\n",
    "\n",
    "                    # Step 1: Meta-analysis of pairwise cell type differences within each SNP for a feature\n",
    "\n",
    "                    # Calculate absolute values of effect sizes\n",
    "                    absolute_effect_sizes <- abs(as.numeric(effect_sizes_condition))\n",
    "\n",
    "                    # Calculate standard errors (SE) of pairwise cell type differences\n",
    "                    pairwise_standard_errors <- as.numeric(se_values_condition)\n",
    "\n",
    "                    # Perform meta-analysis\n",
    "                    meta_result <- rma(yi = absolute_effect_sizes, sei = pairwise_standard_errors, method = \"${meta_method}\")\n",
    "\n",
    "\n",
    "                    z_scores <- meta_result$b / meta_result$se\n",
    "\n",
    "                    # Create a data table to store the feature scores for the current condition\n",
    "                    feature_scores_condition <- data.table(ZScore = z_scores)\n",
    "\n",
    "                    # Append the feature scores for the current condition to the overall feature_scores data table\n",
    "                    feature_scores <- rbindlist(list(feature_scores, feature_scores_condition))\n",
    "                }\n",
    "                colnames(feature_scores) <- gene\n",
    "                condition_name <- colnames(effect_sizes) %>%\n",
    "                    gsub(\"mean_contrast_\", \"\", .) %>%\n",
    "                    gsub(\"_deviation\", \"\", .)\n",
    "                feature_scores <- feature_scores[, condition := condition_name]\n",
    "                saveRDS(feature_scores, str_c(\"${cwd}\",\"/feature_score_meta/cache/\",gsub(\".rds\",\"_featurescore.rds\",basename(res))))\n",
    "            }\n",
    "            if (is.null(out)) {\n",
    "                out <- feature_scores\n",
    "            } else {\n",
    "                out <- merge(out, feature_scores, by = \"condition\", all = TRUE)\n",
    "            }\n",
    "        }\n",
    "       saveRDS(out, \"${_output}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aboriginal-scholar",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## feature_score_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "favorite-hearts",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# compute feature score from contrast results\n",
    "[feature_score_meta_1]\n",
    "#regions = [x.replace(\"\\\"\",\"\").strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "parameter: plink_path = \"/mnt/mfs/cluster/bin/plink1.9.10/plink\"\n",
    "parameter: bfile_path = \"/mnt/vast/hpc/csg/molecular_phenotype_calling/WashU_genotype/genotype_qc/MAP_Brain-xQTL_Gwas_geno_0.1_maf_0.0005.filtered.\"\n",
    "parameter: extract = \"/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/snp.list\"\n",
    "parameter: window_size = '100'\n",
    "parameter: step_size = '10'\n",
    "parameter: r2_threshold = '0.2'\n",
    "parameter: per_chunk = '100'\n",
    "parameter: LD_prune = \"TRUE\"\n",
    "parameter: downsample_ratio = '1'\n",
    "parameter: meta_method = 'REML'\n",
    "parameter: LDcache = \"/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/\"\n",
    "parameter: contrast_input = [path(x[0]) for x in regions]\n",
    "input: contrast_input, group_by = per_chunk\n",
    "output: f\"{cwd}/feature_score_metaLD/cache/mash_posterior_contrast_featurescore{_index+1}.rds\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '20G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    suppressMessages(library(data.table))\n",
    "    suppressMessages(library(tidyverse))\n",
    "    suppressMessages(library(metafor))\n",
    "    plink_path <- \"${plink_path}\"\n",
    "    bfile_path <- \"${bfile_path}\"\n",
    "    \n",
    "    \n",
    "    indep_pairwise <- paste(\"${window_size}\",\"${step_size}\",\"${r2_threshold}\")\n",
    "    out<-NULL\n",
    "    set.seed(999)\n",
    "        for (res in c(${_input:r,})) try({\n",
    "            chr <- res %>%\n",
    "                basename() %>%\n",
    "                str_extract(., paste0(\"\\\\.(.*?)\\\\.\")) %>%\n",
    "                str_replace_all(\"\\\\.\", \"\")%>%\n",
    "                str_replace_all(\"chr\", \"\")\n",
    "            gene <- basename(res) %>%\n",
    "                stringr::str_split(., \"norminal.cis_long_table.\", simplify = TRUE) %>%\n",
    "                .[, 2] %>%\n",
    "                gsub(\"_posterior_contrast.rds\", \"\", .)\n",
    "            extract <- str_c(\"${LDcache}\",gene,\"_snp.list\")\n",
    "            output <- str_c(\"${LDcache}\",gene,\"_output\")\n",
    "            tmp_contrast_results <- readRDS(res)\n",
    "            #downsample with LD prune\n",
    "            if (\"${LD_prune}\" == \"TRUE\") {\n",
    "                tmp_contrast_results <- readRDS(res) %>% as.matrix()\n",
    "      \n",
    "                #  add the rownames for contrast result, should be deleted with new pipeline\n",
    "                orig_data <- res %>%\n",
    "                    basename() %>%\n",
    "                    gsub(\"_posterior_contrast\", \"\", .) %>%\n",
    "                    paste0(\"/mnt/vast/hpc/csg/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.4/add_pQTL/output/RDS/\", .) %>%\n",
    "                    read_rds() %>%\n",
    "                    .[[\"bhat\"]]\n",
    "                posterior_data <- res %>%\n",
    "                    basename() %>%\n",
    "                    gsub(\"_posterior_contrast\", \".posterior\", .) %>%\n",
    "                    paste0(\"/mnt/vast/hpc/csg/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/MASH_6_celltypes_udr_miss/cache/\", .) %>%\n",
    "                    read_rds()\n",
    "                posterior_mean <- posterior_data$PosteriorMean\n",
    "                posterior_cov <- posterior_data$PosteriorCov\n",
    "                orig_data <- orig_data[, colnames(posterior_mean), drop = FALSE]\n",
    "                # orig_data[is.na(orig_data)]<-0#just for test\n",
    "                orig_data[which(is.nan(orig_data))] <- 0 # just for test\n",
    "                rownames(tmp_contrast_results) <- rownames(posterior_mean)\n",
    "\n",
    "                # save the snps in tmp_contrast_results in bash environment\n",
    "                write.table(rownames(tmp_contrast_results), extract, quote = F)\n",
    "\n",
    "                ## run plink\n",
    "                #system(paste0(\"rm \",output,\"*\"), intern = TRUE)\n",
    "                bfile <- str_c(bfile_path, chr)\n",
    "                command <- paste(plink_path, \"--bfile\", bfile, \"--extract\", extract, \"--indep-pairwise\", indep_pairwise, \"--out\", output)\n",
    "                system(command, intern = TRUE)\n",
    "\n",
    "                # read the output from plink\n",
    "                ld_output <- read.table(str_c(\"${LDcache}\",gene,\"_output.prune.in\"))\n",
    "\n",
    "                # new tmp\n",
    "                tmp_contrast_results <- tmp_contrast_results[ld_output$V1, ] %>% as.data.table()\n",
    "            }\n",
    "            #calculate score with meta analysis\n",
    "            if (tmp_contrast_results %>% select(matches(\"mean_contrast.*deviation\")) %>% ncol() > 0) {\n",
    "                effect_sizes <- tmp_contrast_results %>%\n",
    "                    select(matches(\"mean_contrast.*deviation\")) %>%\n",
    "                    as.matrix()\n",
    "                se_values <- tmp_contrast_results %>%\n",
    "                    select(matches(\"se_contrast.*deviation\")) %>%\n",
    "                    as.matrix()\n",
    "\n",
    "                # Initialize an empty data table to store the feature scores for each condition\n",
    "                feature_scores <- data.table()\n",
    "\n",
    "                # Iterate over each condition\n",
    "                for (i in 1:ncol(effect_sizes)) {\n",
    "                    # Subset the effect sizes and standard errors for the current condition\n",
    "                    effect_sizes_condition <- effect_sizes[, i]\n",
    "                    se_values_condition <- se_values[, i]\n",
    "\n",
    "                    # Step 1: Meta-analysis of pairwise cell type differences within each SNP for a feature\n",
    "\n",
    "                    # Calculate absolute values of effect sizes\n",
    "                    absolute_effect_sizes <- abs(as.numeric(effect_sizes_condition))\n",
    "\n",
    "                    # Calculate standard errors (SE) of pairwise cell type differences\n",
    "                    pairwise_standard_errors <- as.numeric(se_values_condition)\n",
    "\n",
    "                    # Perform meta-analysis\n",
    "                    meta_result <- rma(yi = absolute_effect_sizes, sei = pairwise_standard_errors, method = \"${meta_method}\")\n",
    "\n",
    "                    z_scores <- meta_result$b / meta_result$se\n",
    "\n",
    "                    # Create a data table to store the feature scores for the current condition\n",
    "                    feature_scores_condition <- data.table(ZScore = z_scores)\n",
    "\n",
    "                    # Append the feature scores for the current condition to the overall feature_scores data table\n",
    "                    feature_scores <- rbindlist(list(feature_scores, feature_scores_condition))\n",
    "                }\n",
    "                colnames(feature_scores) <- gene\n",
    "                condition_name <- colnames(effect_sizes) %>%\n",
    "                    gsub(\"mean_contrast_\", \"\", .) %>%\n",
    "                    gsub(\"_deviation\", \"\", .)\n",
    "                feature_scores <- feature_scores[, condition := condition_name]\n",
    "                saveRDS(feature_scores, str_c(\"${_output:d}\",\"/\",gsub(\".rds\",\"_featurescore.rds\",basename(res))))\n",
    "            }\n",
    "            if (is.null(out)) {\n",
    "                out <- feature_scores\n",
    "            } else {\n",
    "                out <- merge(out, feature_scores, by = \"condition\", all = TRUE)\n",
    "            }\n",
    "        })\n",
    "       saveRDS(out, \"${_output}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "super-spokesman",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# merge the feature score from contrast results\n",
    "[feature_score_meta_2]\n",
    "input: group_by = \"all\"\n",
    "output: f\"{cwd}/feature_score_metaLD/posterior_feature_score_sum.csv\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '10G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    library(dplyr)\n",
    "    library(tidyverse)\n",
    "    out <- NULL\n",
    "\n",
    "    all.list <- stringr::str_split(\"${_input}\", \" \", simplify = T)\n",
    "    for (i in all.list) {\n",
    "        feature_scores <- readRDS(i)\n",
    "        # Print the feature scores for each condition\n",
    "        if (!is.null(feature_scores)) {\n",
    "            if (is.null(out)) {\n",
    "                out <- feature_scores\n",
    "            } else {\n",
    "                out <- merge(out, feature_scores, by = \"condition\", all = TRUE)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "      out<-t(out)\n",
    "      #out<-out[-1,]\n",
    "    write.csv(out, \"${_output}\",col.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "correct-separation",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# compute feature score from contrast results\n",
    "#[feature_score_1]\n",
    "#regions = [x.replace(\"\\\"\",\"\").strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "input: regions, group_by = 1\n",
    "output: f\"{cwd}/feature_score/{_input:bn}_posterior_contrast_featurescore.rds\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '20G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    suppressMessages(library(data.table))\n",
    "    suppressMessages(library(tidyverse))\n",
    "    suppressMessages(library(metafor))\n",
    "\n",
    "    gene <- basename(\"${_input}\") %>%\n",
    "        stringr::str_split(., \"norminal.cis_long_table.\", simplify = TRUE) %>%\n",
    "        .[, 2] %>%\n",
    "        gsub(\"_posterior_contrast.rds\", \"\", .)\n",
    "\n",
    "    tmp_contrast_results <- readRDS(\"${_input}\")\n",
    "    if (tmp_contrast_results %>% select(matches(\"mean_contrast.*deviation\")) %>% ncol() > 0) {\n",
    "        feature_out<-list()\n",
    "        effect_sizes <- tmp_contrast_results %>%\n",
    "            select(matches(\"mean_contrast.*deviation\")) %>%\n",
    "            as.matrix()\n",
    "        se_values <- tmp_contrast_results %>%\n",
    "            select(matches(\"se_contrast.*deviation\")) %>%\n",
    "            as.matrix()\n",
    "\n",
    "        # Initialize an empty data table to store the feature scores for each condition\n",
    "        feature_scores <- data.table()\n",
    "        avg_snp_diff <- data.table()\n",
    "        avg_diff <- data.table()\n",
    "\n",
    "        # Iterate over each condition\n",
    "        for (i in 1:ncol(effect_sizes)) {\n",
    "            # Subset the effect sizes and standard errors for the current condition\n",
    "            effect_sizes_condition <- effect_sizes[, i]\n",
    "            se_values_condition <- se_values[, i]\n",
    "\n",
    "            # Step 1: Meta-analysis of pairwise cell type differences within each SNP for a feature\n",
    "\n",
    "            # Calculate absolute values of effect sizes\n",
    "            absolute_effect_sizes <- abs(as.numeric(effect_sizes_condition))\n",
    "\n",
    "            # Calculate standard errors (SE) of pairwise cell type differences\n",
    "            pairwise_standard_errors <- as.numeric(se_values_condition)\n",
    "\n",
    "            # Perform meta-analysis\n",
    "            meta_result <- rma(yi = absolute_effect_sizes, sei = pairwise_standard_errors, method = \"${meta_method}\")\n",
    "\n",
    "            # Get meta-analysis estimate of the average cell type difference\n",
    "            average_difference <- meta_result$b\n",
    "            avg_diff <- rbindlist(list(avg_diff, as.data.table(average_difference)))\n",
    "\n",
    "            # Step 2: Second round of meta-analysis on SNP-level meta-analyzed effect size and standard error across all the SNPs in the gene\n",
    "\n",
    "            # Calculate effect size and standard error across all SNPs in the gene\n",
    "            gene_effect_size <- mean(as.numeric(effect_sizes_condition))\n",
    "            gene_standard_error <- sqrt(sum(as.numeric(se_values_condition)^2))\n",
    "\n",
    "            # Perform meta-analysis\n",
    "            meta_result_gene <- rma(yi = gene_effect_size, sei = gene_standard_error, method = \"${meta_method}\")\n",
    "\n",
    "            # Get meta-analysis estimate of the average cell type difference in the average SNP\n",
    "            average_snp_difference <- meta_result_gene$b\n",
    "            avg_snp_diff <- rbindlist(list(avg_snp_diff, as.data.table(average_snp_difference)))\n",
    "\n",
    "            # Calculate z-scores for each feature\n",
    "            z_scores <- (average_difference - average_snp_difference) / meta_result_gene$se\n",
    "\n",
    "            # Create a data table to store the feature scores for the current condition\n",
    "            feature_scores_condition <- data.table(Feature = rownames(effect_sizes), ZScore = z_scores)\n",
    "\n",
    "            # Append the feature scores for the current condition to the overall feature_scores data table\n",
    "            feature_scores <- rbindlist(list(feature_scores, feature_scores_condition))\n",
    "        }\n",
    "        colnames(avg_diff)<-colnames(avg_snp_diff)<-colnames(feature_scores)<-gene\n",
    "        condition_name<-colnames(effect_sizes)%>%gsub(\"mean_contrast_\",\"\",.)%>%gsub(\"_deviation\",\"\",.)\n",
    "        avg_diff<-avg_diff[,condition := condition_name]\n",
    "        avg_snp_diff<-avg_snp_diff[,condition := condition_name]\n",
    "        feature_scores<-feature_scores[,condition := condition_name]\n",
    "        feature_out$avg_diff<-avg_diff\n",
    "        feature_out$avg_snp_diff<-avg_snp_diff\n",
    "        feature_out$feature_scores<-feature_scores\n",
    "    } else {\n",
    "        feature_out <- NULL\n",
    "    }\n",
    "    \n",
    "    saveRDS(feature_out,\"${_output}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "appropriate-relay",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# merge the feature score from contrast results\n",
    "#[feature_score_2]\n",
    "input: group_by = \"all\"\n",
    "output: f\"{cwd}/feature_score/posterior_feature_score_sum.csv\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '10G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    library(dplyr)\n",
    "    library(tidyverse)\n",
    "    out <- list()\n",
    "\n",
    "    all.list <- stringr::str_split(\"${_input}\", \" \", simplify = T)\n",
    "    for (i in all.list) {\n",
    "        feature_scores <- readRDS(i)\n",
    "        # Print the feature scores for each condition\n",
    "        if (!is.null(feature_scores)) {\n",
    "           if (is.null(out$zscore)) {\n",
    "                out$avg_diff<-feature_scores$avg_diff\n",
    "                out$avg_snp_diff<-feature_scores$avg_snp_diff\n",
    "                out$zscore <- feature_scores$feature_scores\n",
    "              } else {\n",
    "                out$avg_diff<- merge(out$avg_diff, feature_scores$feature_scores, by = \"condition\", all = TRUE)\n",
    "                out$avg_snp_diff<- merge(out$avg_snp_diff, feature_scores$feature_scores, by = \"condition\", all = TRUE)\n",
    "                out$zscore <- merge(out$zscore, feature_scores$feature_scores, by = \"condition\", all = TRUE)\n",
    "              }\n",
    "        }\n",
    "    }\n",
    "    write.csv(out, \"${_output}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "integrated-bread",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "painted-municipality",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# compute feature score from contrast results with pairwise data\n",
    "#[feature_score_pw_1]\n",
    "#regions = [x.replace(\"\\\"\",\"\").strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "parameter: job_size = 100\n",
    "parameter: significance_threshold = 1 \n",
    "input: regions\n",
    "output: f\"{cwd}/feature_score_{meta_method}/cache/{_input:bn}_posterior_contrast_featurescore.rds\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '60G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "        suppressMessages(library(data.table))\n",
    "        suppressMessages(library(tidyverse))\n",
    "        suppressMessages(library(metafor))\n",
    "\n",
    "        gene <- basename(\"${_input}\") %>%\n",
    "            stringr::str_split(., \"norminal.cis_long_table.\", simplify = TRUE) %>%\n",
    "            .[, 2] %>%\n",
    "            gsub(\"_posterior_contrast.rds\", \"\", .)\n",
    "\n",
    "        tmp_contrast_results <- readRDS(\"${_input}\")\n",
    "        if (${significance_threshold} < 1) tmp_contrast_results <- tmp_contrast_results%>%select(matches(\"p_contrast_.*_vs_*\"))%>% .[rowSums(. < ${significance_threshold}) >0, ]## reduce calculation\n",
    "  \n",
    "        if (tmp_contrast_results %>% select(matches(\"mean_contrast.*_vs_*\")) %>% ncol() > 0) {\n",
    "            effect_sizes <- tmp_contrast_results %>%\n",
    "                select(matches(\"mean_contrast.*_vs_*\")) %>%\n",
    "                as.matrix()\n",
    "            se_values <- tmp_contrast_results %>%\n",
    "                select(matches(\"se_contrast.*_vs_*\")) %>%\n",
    "                as.matrix()\n",
    "        \n",
    "            conditions <- colnames(effect_sizes) %>%\n",
    "                sub(\"mean_contrast_\", \"\", .) %>%\n",
    "                unique()\n",
    "          \n",
    "            cells <- c(sub(\"_vs_.*\", \"\", conditions), sub(\".*_vs_\", \"\", conditions)) %>% unique()\n",
    "\n",
    "            df <- data.frame()\n",
    "            for (cell in cells) {\n",
    "                print(cell)\n",
    "                cell.b <- effect_sizes[, grep(cell, colnames(effect_sizes)),drop=F]\n",
    "                cell.se <- se_values[, grep(cell, colnames(se_values)),drop=F]\n",
    "\n",
    "                # step1\n",
    "                avg_diff <- c()\n",
    "                avg_se <- c()\n",
    "                for (i in 1:nrow(cell.b)) {\n",
    "                    # Subset the effect sizes and standard errors for the current condition\n",
    "                    effect_sizes_condition <- cell.b[i, ]\n",
    "                    se_values_condition <- cell.se[i, ]\n",
    "\n",
    "                    # Calculate absolute values of effect sizes\n",
    "                    absolute_effect_sizes <- abs(as.numeric(effect_sizes_condition))\n",
    "\n",
    "                    # Calculate standard errors (SE) of pairwise cell type differences\n",
    "                    pairwise_standard_errors <- as.numeric(se_values_condition)\n",
    "\n",
    "                    # Perform meta-analysis using random effects model (REML)\n",
    "                    meta_result <- rma(yi = absolute_effect_sizes, sei = pairwise_standard_errors, method = \"${meta_method}\")\n",
    "\n",
    "                    # Get meta-analysis estimate of the average cell type difference\n",
    "                    avg_diff <- c(avg_diff, meta_result$b)\n",
    "                    avg_se <- c(avg_se, meta_result$se)\n",
    "                    # Calculate z-scores for each SNP\n",
    "                }\n",
    "                # step2\n",
    "                meta_result_con <- rma(yi = avg_diff, sei = avg_se, method = \"${meta_method}\")\n",
    "                # Calculate z-scores for each condition\n",
    "                z_scores <- meta_result_con$b / meta_result_con$se\n",
    "                df[cell, 1] <- z_scores\n",
    "            }\n",
    "            colnames(df) <- gene\n",
    "        } else {\n",
    "            df <- NULL\n",
    "        }\n",
    "\n",
    "        saveRDS(df, \"${_output}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "voluntary-driving",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# merge the feature score from contrast results with pairwise data\n",
    "#[feature_score_pw_2]\n",
    "input: group_by = \"all\"\n",
    "output: f\"{cwd}/feature_score_{meta_method}/posterior_feature_score_sum.csv\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '10G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    library(dplyr)\n",
    "    library(tidyverse)\n",
    "    out <- list()\n",
    "\n",
    "    all.list <- stringr::str_split(\"${_input}\", \" \", simplify = T)\n",
    "    for (i in all.list) {\n",
    "        feature_scores <- readRDS(i)\n",
    "        # Print the feature scores for each condition\n",
    "        if (!is.null(feature_scores)) {\n",
    "           if (is.null(out$zscore)) {\n",
    "                out <- t(feature_scores)%>%as.data.table\n",
    "              } else {\n",
    "                out <- rbindlist(list(out, t(feature_scores)%>%as.data.table), use.names = TRUE)\n",
    "              }\n",
    "        }\n",
    "    }\n",
    "    write.csv(out, \"${_output}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "charged-compilation",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## feature_pval_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "american-texture",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# compute feature score from contrast results\n",
    "[feature_pval_pair_1]\n",
    "#regions = [x.replace(\"\\\"\",\"\").strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "parameter: plink_path = \"/mnt/mfs/cluster/bin/plink1.9.10/plink\"\n",
    "parameter: bfile_path = \"/mnt/vast/hpc/csg/molecular_phenotype_calling/WashU_genotype/genotype_qc/MAP_Brain-xQTL_Gwas_geno_0.1_maf_0.0005.filtered.\"\n",
    "parameter: extract = \"/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/snp.list\"\n",
    "parameter: window_size = '100'\n",
    "parameter: step_size = '10'\n",
    "parameter: r2_threshold = '0.2'\n",
    "parameter: per_chunk = '100'\n",
    "parameter: LD_prune = \"TRUE\"\n",
    "parameter: downsample_ratio = '1'\n",
    "parameter: meta_method = 'REML'\n",
    "parameter: se_cutoff = '1E-03'\n",
    "parameter: LDcache = \"/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/\"\n",
    "parameter: contrast_input = [path(x[0]) for x in regions]\n",
    "input: contrast_input, group_by = per_chunk\n",
    "output: f\"{cwd}/feature_pval_pair/cache/mash_posterior_contrast_featurescore{_index+1}.rds\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = mem, tags = f'{_output:bn}'  \n",
    "\n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    suppressMessages(library(data.table))\n",
    "    suppressMessages(library(tidyverse))\n",
    "    suppressMessages(library(metafor))\n",
    "    out<-NULL\n",
    "    set.seed(999)\n",
    "        for (res in c(${_input:r,})) {\n",
    "        gene <- basename(res) %>%\n",
    "            stringr::str_split(., \"norminal.cis_long_table.\", simplify = TRUE) %>%\n",
    "            .[, 2] %>%\n",
    "            gsub(\"_posterior_contrast.rds\", \"\", .)\n",
    "        print(gene)\n",
    "        #res <- results[grep(str_c(\"norminal.cis_long_table.\", gene, \"_posterior_contrast.rds\"), results)]\n",
    "        tmp_contrast_results <- readRDS(res)\n",
    "  \n",
    "        ###  add the rownames for contrast result, should be deleted with new pipeline\n",
    "        orig_data <- res %>%\n",
    "            basename() %>%\n",
    "            gsub(\"_posterior_contrast\", \"\", .) %>%\n",
    "            paste0(\"/mnt/vast/hpc/csg/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.4/add_pQTL/output/RDS/\", .) %>%\n",
    "            read_rds() %>%\n",
    "            .[[\"bhat\"]]\n",
    "        posterior_data <- res %>%\n",
    "            basename() %>%\n",
    "            gsub(\"_posterior_contrast\", \".posterior\", .) %>%\n",
    "            paste0(\"/mnt/vast/hpc/csg/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/MASH_6_celltypes_udr_miss/cache/\", .) %>%\n",
    "            read_rds()\n",
    "        posterior_mean <- posterior_data$PosteriorMean\n",
    "        posterior_cov <- posterior_data$PosteriorCov\n",
    "        orig_data <- orig_data[, colnames(posterior_mean), drop = FALSE]\n",
    "        # orig_data[is.na(orig_data)]<-0#just for test\n",
    "        orig_data[which(is.nan(orig_data))] <- 0 # just for test\n",
    "        rownames(tmp_contrast_results) <- rownames(posterior_mean)\n",
    "        ##\n",
    "  \n",
    "        # read the output from plink\n",
    "        ld_output <- read.table(str_c(\"${LDcache}\",gene,\"_output.prune.in\"))\n",
    "        # new tmp\n",
    "        tmp_contrast_results <- tmp_contrast_results[ld_output$V1, ] %>% as.data.table(keep.rownames = T)\n",
    "\n",
    "        effect_sizes <- tmp_contrast_results %>%\n",
    "            select(matches(\"mean_contrast.*_vs_*\")) %>%\n",
    "            as.matrix()\n",
    "        se_values <- tmp_contrast_results %>%\n",
    "            select(matches(\"se_contrast.*_vs_*\")) %>%\n",
    "            as.matrix()\n",
    "\n",
    "        conditions <- colnames(effect_sizes) %>%\n",
    "            sub(\"mean_contrast_\", \"\", .) %>%\n",
    "            unique()\n",
    "        cells <- c(sub(\"_vs_.*\", \"\", conditions), sub(\".*_vs_\", \"\", conditions)) %>% unique()\n",
    "\n",
    "        # perform meta analysis per cell (different cell type)\n",
    "        df <- data.table()\n",
    "        for (cell in cells) {\n",
    "            #print(cell)\n",
    "            cell.b <- effect_sizes[, grep(cell, colnames(effect_sizes)),drop=F]\n",
    "            cell.se <- se_values[, grep(cell, colnames(se_values)),drop=F]\n",
    "\n",
    "            feature_pvals <- data.table()\n",
    "            for (i in 1:ncol(cell.b)) {\n",
    "                # Subset the effect sizes and standard errors for the current condition\n",
    "                effect_sizes_condition <- cell.b[, i]\n",
    "                se_values_condition <- cell.b[, i]\n",
    "                #delete the snps with super small se\n",
    "                effect_sizes_condition<-effect_sizes_condition[which(se_values_condition>as.numeric(\"${se_cutoff}\"))]\n",
    "                se_values_condition<-se_values_condition[which(se_values_condition>as.numeric(\"${se_cutoff}\"))]\n",
    "                # Step 1: Meta-analysis of pairwise cell type differences within each SNP for a feature\n",
    "\n",
    "                # Calculate absolute values of effect sizes\n",
    "                absolute_effect_sizes <- abs(as.numeric(effect_sizes_condition))\n",
    "\n",
    "                # Calculate standard errors (SE) of pairwise cell type differences\n",
    "                pairwise_standard_errors <- as.numeric(se_values_condition)\n",
    "\n",
    "                # Perform meta-analysis\n",
    "                meta_result <- rma(yi = absolute_effect_sizes, sei = pairwise_standard_errors, method = \"REML\")\n",
    "\n",
    "\n",
    "                # Create a data table to store the feature scores for the current condition\n",
    "                feature_pval_condition <- data.table(pavlue = meta_result$pval)\n",
    "\n",
    "                # Append the feature scores for the current condition to the overall feature_scores data table\n",
    "                feature_pvals <- rbindlist(list(feature_pvals, feature_pval_condition))\n",
    "            }\n",
    "            colnames(feature_pvals) <- gene\n",
    "            condition_name <- colnames(cell.b) %>%\n",
    "                gsub(\"mean_contrast_\", \"\", .)\n",
    "            feature_pvals <- feature_pvals[, condition := condition_name]\n",
    "            df <- rbindlist(list(df, feature_pvals))\n",
    "            saveRDS(df, str_c(\"${_output:d}\",\"/\",gsub(\".rds\",\"_featurescore_pw.rds\",basename(res))))\n",
    "        }\n",
    "    \n",
    "    if (is.null(out)) {\n",
    "        out <- df\n",
    "    } else {\n",
    "        out <- merge(out, df, by = \"condition\", all = TRUE, allow.cartesian=TRUE)\n",
    "    }\n",
    "    }\n",
    "    saveRDS(out, \"${_output}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "anticipated-respondent",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# merge the feature score from contrast results\n",
    "[feature_pval_pair_2]\n",
    "input: group_by = \"all\"\n",
    "output: f\"{cwd}/feature_pval_pair/posterior_feature_score_sum.csv\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '10G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    library(dplyr)\n",
    "    library(tidyverse)\n",
    "    out <- NULL\n",
    "\n",
    "    all.list <- stringr::str_split(\"${_input}\", \" \", simplify = T)\n",
    "    for (i in all.list) {\n",
    "        feature_scores <- readRDS(i)\n",
    "        # Print the feature scores for each condition\n",
    "        if (!is.null(feature_scores)) {\n",
    "            if (is.null(out)) {\n",
    "                out <- feature_scores\n",
    "            } else {\n",
    "                out <- merge(out, feature_scores, by = \"condition\", all = TRUE)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "      out<-t(out)\n",
    "      #out<-out[-1,]\n",
    "    write.csv(out, \"${_output}\",col.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fatal-activation",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## feature_score_finemap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "demanding-garbage",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# compute feature score from contrast results with eQTL and pQTL finemapped signals\n",
    "[feature_score_finemap_1]\n",
    "#regions = [x.replace(\"\\\"\",\"\").strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "parameter: per_chunk = '100'\n",
    "parameter: pfine_path = \"/mnt/vast/hpc/csg/molecular_phenotype_calling/QTL_fine_mapping/pqtl.all_variants.tsv\"\n",
    "parameter: efine_path = \"/mnt/vast/hpc/csg/hs3393/susie_eqtl/Result_new/tsv/\"\n",
    "parameter: efine_suffix = \".unisusie.fit.variant.tsv\"\n",
    "parameter: gene_ref_path = \"/mnt/vast/hpc/csg/molecular_phenotype_calling/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.region_list\"\n",
    "parameter: contrast_input = [path(x[0]) for x in regions]\n",
    "input: contrast_input, group_by = per_chunk\n",
    "output: f\"{cwd}/feature_score_finemap/cache/mash_posterior_contrast_featurescore{_index+1}.rds\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '40G', tags = f'{_output:bn}'  \n",
    "\n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "        suppressMessages(library(data.table))\n",
    "        suppressMessages(library(tidyverse))\n",
    "        score_from_cs <- function(fine.file, contrast_results, con) { # find top signal in each CS\n",
    "            CSs <- unique(fine.file$cs_order) %>% .[. != 0]\n",
    "            max.cs.df <- NULL\n",
    "            # could be more than 1 CS in one file, pick all the most significant snps from them\n",
    "            for (cs in CSs) {\n",
    "                tmp <- fine.file[fine.file$cs_order == cs, ]\n",
    "                max.cs <- tmp[tmp$pip == max(tmp$pip), ]\n",
    "                max.cs.df <- rbind(max.cs.df, max.cs)\n",
    "            }\n",
    "            # pick the snp with the most significant pvalue in contrast result\n",
    "            # some strongest signal in CS are not present in the sumstats file because of the difference between cis and tad, which won't be a problem in new generated data.\n",
    "            # here I just picked the overlap (may not be the most correct way)\n",
    "            snps <- intersect(rownames(contrast_results), max.cs.df$variants)\n",
    "            if (length(snps) > 0) {\n",
    "                # if only 2 groups and no deviation result\n",
    "                if (contrast_results %>% as.data.frame() %>% select(matches(str_c(\"p_contrast_\", con, \"_deviation\"))) %>% ncol() > 0) {\n",
    "                    contrast_p <- contrast_results[snps, str_c(\"p_contrast_\", con, \"_deviation\"), drop = F]\n",
    "                    # in case some snps may have more than 1 most significant p value\n",
    "                    max.snp <- contrast_p[contrast_p[, 1] == max(contrast_p[, 1]), , drop = F] %>% rownames()\n",
    "                    # compute zscore with beta and se\n",
    "                    score <- abs(contrast_results[max.snp, str_c(\"mean_contrast_\", con, \"_deviation\")]) / contrast_results[max.snp, str_c(\"se_contrast_\", con, \"_deviation\")]\n",
    "                    score <- max(score)\n",
    "                } else if (contrast_results %>% as.data.frame() %>% select(matches(\"p_contrast.*_vs_*\")) %>% ncol() == 1) {\n",
    "                    contrast_p <- contrast_results %>%\n",
    "                        as.data.frame() %>%\n",
    "                        select(matches(\"p_contrast.*_vs_*\")) %>%\n",
    "                        .[snps, , drop = F]\n",
    "                    # in case some snps may have more than 1 most significant p value\n",
    "                    max.snp <- contrast_p[contrast_p[, 1] == max(contrast_p[, 1]), , drop = F] %>% rownames()\n",
    "                    # compute zscore with beta and se\n",
    "                    score <- abs(contrast_results %>% as.data.frame() %>% select(matches(\"mean_contrast.*_vs_*\")) %>% .[max.snp, ]) / contrast_results %>%\n",
    "                        as.data.frame() %>%\n",
    "                        select(matches(\"se_contrast.*_vs_*\")) %>%\n",
    "                        .[max.snp, ]\n",
    "                    score <- max(score)\n",
    "                }\n",
    "            } else {\n",
    "                score <- NA\n",
    "            }\n",
    "\n",
    "            return(score)\n",
    "        }\n",
    "\n",
    "\n",
    "        # read fine mapped pQTL file\n",
    "        pfine.mapped.result <- fread(\"${pfine_path}\")\n",
    "        pfine.mapped.result$Gene <- stringr::str_split(pfine.mapped.result$molecular_trait_id, \"_\", simplify = T)[, 2]\n",
    "        gene.ref <- read.table(\"${gene_ref_path}\")\n",
    "\n",
    "\n",
    "        # results <- list.files(path = \"/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/MASH_6_celltypes_udr_miss/contrast\", \"posterior_contrast.rds\", full.names = T)\n",
    "        out <- data.table()\n",
    "        # for (res in results[1:10]) {\n",
    "        for (res in c(${_input:r,})) {\n",
    "            df <- data.frame()\n",
    "\n",
    "            tmp_contrast_results <- readRDS(res) %>% as.matrix()\n",
    "\n",
    "\n",
    "            #  add the rownames for contrast result, should be deleted with new pipeline\n",
    "            orig_data <- res %>%\n",
    "                basename() %>%\n",
    "                gsub(\"_posterior_contrast\", \"\", .) %>%\n",
    "                paste0(\"/mnt/vast/hpc/csg/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.4/add_pQTL/output/RDS/\", .) %>%\n",
    "                read_rds() %>%\n",
    "                .[[\"bhat\"]]\n",
    "            posterior_data <- res %>%\n",
    "                basename() %>%\n",
    "                gsub(\"_posterior_contrast\", \".posterior\", .) %>%\n",
    "                paste0(\"/mnt/vast/hpc/csg/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/MASH_6_celltypes_udr_miss/cache/\", .) %>%\n",
    "                read_rds()\n",
    "            posterior_mean <- posterior_data$PosteriorMean\n",
    "            posterior_cov <- posterior_data$PosteriorCov\n",
    "            orig_data <- orig_data[, colnames(posterior_mean), drop = FALSE]\n",
    "            orig_data[which(is.nan(orig_data))] <- 0 # just for test\n",
    "            rownames(tmp_contrast_results) <- rownames(posterior_mean)\n",
    "\n",
    "            #\n",
    "\n",
    "\n",
    "            gene <- basename(res) %>%\n",
    "                stringr::str_split(., \"norminal.cis_long_table.\", simplify = TRUE) %>%\n",
    "                .[, 2] %>%\n",
    "                gsub(\"_posterior_contrast.rds\", \"\", .)\n",
    "\n",
    "\n",
    "            ensemble <- gene.ref[gene.ref$V5 == gene, ]$V4\n",
    "\n",
    "            efine.mapped.result <- list.files(path = \"${efine_path}\", pattern = paste0(ensemble, \"${efine_suffix}\"), full.names = T)\n",
    "\n",
    "            pfine.file <- pfine.mapped.result[pfine.mapped.result$Gene == gene & pfine.mapped.result$cs_order > 0, ]\n",
    "\n",
    "            if (nrow(pfine.file) > 0) {\n",
    "                message(\"Extracting signal from pQTL in \", gene)\n",
    "\n",
    "                df[gene, \"DLPFC_pQTL\"] <- score_from_cs(fine.file = pfine.file, contrast_results = tmp_contrast_results, con = \"DLPFC_pQTL\")\n",
    "            }\n",
    "\n",
    "\n",
    "            if (length(efine.mapped.result) > 0) {\n",
    "                print(gene)\n",
    "                fine.conditions <- basename(efine.mapped.result) %>%\n",
    "                    sub(\"demo.\", \"\", .) %>%\n",
    "                    sub(paste0(\".\", ensemble, \"${efine_suffix}\"), \"\", .) %>%\n",
    "                    .[. != \"ALL\"] %>%\n",
    "                    .[. != \"End\"]\n",
    "\n",
    "                for (con in fine.conditions) {\n",
    "                    # print(con)\n",
    "                    # con=\"Ast\"\n",
    "                    con.file <- efine.mapped.result[grep(con, efine.mapped.result)] %>% fread()\n",
    "                    con.fine.file <- con.file[con.file$cs_order > 0, ]\n",
    "                    if (nrow(con.fine.file) > 0) {\n",
    "                        df[gene, con] <- score_from_cs(fine.file = con.fine.file, contrast_results = tmp_contrast_results, con = con)\n",
    "                    }\n",
    "                }\n",
    "            }\n",
    "            saveRDS(df, str_c(\"${cwd}\",\"/feature_score_finemap/cache/\",gsub(\".rds\",\"_featurescore.rds\",basename(res))))\n",
    "            out <- rbindlist(list(out, as.data.table(df, keep.rownames = TRUE)), use.names = TRUE, fill = TRUE)\n",
    "        }\n",
    "        saveRDS(out, \"${_output}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "incorporated-driving",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# merge the feature score from contrast results with eQTL and pQTL finemapped signals\n",
    "[feature_score_finemap_2]\n",
    "input: group_by = \"all\"\n",
    "output: f\"{cwd}/feature_score_finemap/posterior_feature_score_sum.csv\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '10G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    suppressMessages(library(data.table))\n",
    "    suppressMessages(library(tidyverse))\n",
    "    \n",
    "    out <- data.table()\n",
    "\n",
    "    all.list <- stringr::str_split(\"${_input}\", \" \", simplify = T)\n",
    "    for (i in all.list) {\n",
    "        feature_scores <- readRDS(i)\n",
    "        # Print the feature scores for each condition\n",
    "        if (!is.null(feature_scores)) {\n",
    "           if (is.null(out)) {\n",
    "                out <- as.data.table(feature_scores, keep.rownames = TRUE)\n",
    "              } else {\n",
    "                out <-  rbindlist(list(out, as.data.table(feature_scores, keep.rownames = TRUE)), use.names = TRUE, fill = TRUE)\n",
    "              }\n",
    "        }\n",
    "    }\n",
    "    write.csv(out, \"${_output}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "operating-jackson",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## feature_score_nsig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "desperate-stylus",
   "metadata": {
    "kernel": "Bash",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1]+  Done                    nohup sos run /home/rf2872/codes/xqtl-pipeline/pipeline/contrast_feature_score.ipynb feature_score_nsig --analysis_unit <(ls ~/Work/Multivariate/MASH/From_SuSiE/2023.5_new/ep_MiGA/MASH_6_celltypes_udr_miss/contrast_merge/*posterior_contrast.rds) --cwd MASH_6_celltypes_udr_miss/contrast_merge -s force -J 200 -q csg -c ~/test/csg.yml &> mash_contrast_score_nsig.log\n",
      "[1] 45008\n"
     ]
    }
   ],
   "source": [
    "cd ~/Work/Multivariate/MASH/From_SuSiE/2023.5_new/ep_MiGA/\n",
    "nohup sos run /home/rf2872/codes/xqtl-pipeline/pipeline/contrast_feature_score.ipynb feature_score_nsig    \\\n",
    "--analysis_unit   <(ls ~/Work/Multivariate/MASH/From_SuSiE/2023.5_new/ep_MiGA/MASH_6_celltypes_udr_miss/contrast_merge/*posterior_contrast.rds) \\\n",
    "--cwd MASH_6_celltypes_udr_miss/contrast_merge \\\n",
    "-s force -J 200 -q csg -c ~/test/csg.yml  &> mash_contrast_score_nsig.log &  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "elementary-asset",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1]+  Done                    nohup sos run /home/rf2872/codes/xqtl-pipeline/pipeline/contrast_feature_score.ipynb feature_score_nsig --analysis_unit <(ls ~/Work/Multivariate/MASH/From_SuSiE/2023.5_new/ep_MiGA/MASH_6_celltypes_udr_miss/contrast_merge/*posterior_contrast.rds) --cwd MASH_6_celltypes_udr_miss/contrast_merge -s force -J 200 -q csg -c ~/test/csg.yml &> mash_contrast_score_nsig.log\n",
      "[1] 48608\n"
     ]
    }
   ],
   "source": [
    "cd ~/Work/Multivariate/MASH/From_SuSiE/2023.5_new/ep_MiGA/\n",
    "nohup sos run /home/rf2872/codes/xqtl-pipeline/pipeline/contrast_feature_score.ipynb feature_score_nsig    \\\n",
    "--analysis_unit   <(ls ~/Work/Multivariate/MASH/From_SuSiE/2023.5_new/ep_MiGA/MASH_6_celltypes_udr_miss/contrast_merge_non/*posterior_contrast.rds) \\\n",
    "--cwd MASH_6_celltypes_udr_miss/contrast_merge_non \\\n",
    "-s build -J 200 -q csg -c ~/test/csg.yml  &> mash_contrast_score_non_nsig.log &  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "surprising-description",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# compute feature score from contrast results\n",
    "[feature_score_nsig_1]\n",
    "#regions = [x.replace(\"\\\"\",\"\").strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "parameter: per_chunk = '100'\n",
    "parameter: p_cut=0.00001\n",
    "parameter: contrast_input = [path(x[0]) for x in regions]\n",
    "input: contrast_input, group_by = per_chunk\n",
    "output: f\"{cwd}/feature_score_nsig/cache/mash_posterior_contrast_featurescore_nsig{_index+1}.rds\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '20G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    suppressMessages(library(data.table))\n",
    "    suppressMessages(library(tidyverse))\n",
    "    suppressMessages(library(metafor))\n",
    "    p_cut <-${p_cut}%>%as.numeric\n",
    "    out<-NULL\n",
    "    set.seed(999)\n",
    "        for (res in c(${_input:r,})){\n",
    "            try({\n",
    "                #res <- readRDS(res)\n",
    "                gene <- basename(res) %>%\n",
    "                    stringr::str_split(., \"norminal.cis_long_table.\", simplify = TRUE) %>%\n",
    "                    .[, 2] %>%\n",
    "                    gsub(\"_posterior_contrast.rds\", \"\", .)\n",
    "                print(gene)\n",
    "                tmp_contrast_results <- readRDS(res) %>%\n",
    "                    select(matches(\"p_contrast_.*deviation\")) %>%\n",
    "                    as.matrix()\n",
    "                # Iterate over each condition\n",
    "                feature.out <- data.table()\n",
    "                for (i in 1:ncol(tmp_contrast_results)) {\n",
    "                    feature.out <- feature.out[, c(\"n_sig_snp\", \"n_snp\", \"condition\") := list(colSums(tmp_contrast_results < p_cut, na.rm = TRUE), colSums(tmp_contrast_results, na.rm = TRUE), colnames(tmp_contrast_results) %>% gsub(\"p_contrast_\", \"\", .) %>% gsub(\"_deviation\", \"\", .))]\n",
    "                    feature.out[, gene] <- feature.out$n_sig_snp / feature.out$n_snp\n",
    "                }\n",
    "                saveRDS(feature.out, paste0(\"${_output:d}\",\"/\", gsub(\".rds\", \"_n_sig_ratio.rds\", basename(res))))\n",
    "                if (is.null(out)) {\n",
    "                    out <- feature.out %>%\n",
    "                        as.data.frame() %>%\n",
    "                        .[, c(gene, \"condition\")]\n",
    "                } else {\n",
    "                    out <- merge(out, feature.out %>% as.data.frame() %>% .[, c(gene, \"condition\")], by = \"condition\", all = TRUE)\n",
    "                }\n",
    "            })\n",
    "        }\n",
    "       saveRDS(out, \"${_output}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "attempted-roads",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# merge the feature score from contrast results\n",
    "[feature_score_nsig_2]\n",
    "input: group_by = \"all\"\n",
    "output: f\"{cwd}/feature_score_nsig/posterior_feature_score_sum.csv\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '24h',  mem = '10G', tags = f'{_output:bn}'  \n",
    "\n",
    "R: expand = \"${ }\",stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout' \n",
    "    library(dplyr)\n",
    "    library(tidyverse)\n",
    "    out <- NULL\n",
    "\n",
    "    all.list <- stringr::str_split(\"${_input}\", \" \", simplify = T)\n",
    "    for (i in all.list) {\n",
    "        feature_scores <- readRDS(i)\n",
    "        # Print the feature scores for each condition\n",
    "        if (!is.null(feature_scores)) {\n",
    "            if (is.null(out)) {\n",
    "                out <- feature_scores\n",
    "            } else {\n",
    "                out <- merge(out, feature_scores, by = \"condition\", all = TRUE)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "      out<-t(out)\n",
    "      #out<-out[-1,]\n",
    "    write.csv(out, \"${_output}\",col.names=F)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "bash",
     "Bash",
     "#E6EEFF",
     "shell"
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.22.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
